### PR TITLE
fixes #4021 refactor(nimbus): Add 'useCommonForm' w/error clear onChange, use in Overview, Metrics, + Audience forms

### LINKED ISSUE #4021: Nimbus Public name field doesnt validate on blur (create pattern for server-side validation clearing onblur)
https://stage.experimenter.nonprod.dataops.mozgcp.net/experiments/nimbus/aaab/edit/overview

If I set the experiment name to "aaa", i get a validation error saying "Nimbus Experiment with this name already exists". If I change the experiment name to "aaab" and tab or focus out of that field, the error message persists, despite the name now being valid. It appears validation only happens again if I click "Save" or "Next" buttons.

Mozilla_Experimenter

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-614)
┆Issue Number: EXP-614

### PR DESCRIPTION
**Pull request by LZoog**
Because:
* We have a lot of common functionality floating around and need a place to pull from for consistency across all of our forms.
* We want server errors to clear onChange of that field.

This commit:
* Creates useCommonForm hook
* Refactors/makes FormOverview, FormMetrics, and FormAudience consistent using said hook
* Contains additional MetricsForm refactoring for consistency/readability
* Clears per-field submitError onChange of fields for inputs, selects, and multiselects
* Displays submit errors in FormMetrics
* Cleans up some tests

fixes #4021

---

I'll create follow up tickets:
1) to use `useCommonForm` in the Branches page: https://jira.mozilla.com/browse/EXP-781

### CODE REVIEW

File: app/yarn.lock
Comment (LZoog) [inline app/yarn.lock:547]: Dunno why this yarn lock change, but should be fixed once I rebase

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:96]: I'm going to refactor some related stuff in the Metrics form, it's a bit funky.

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.test.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.test.tsx:115]: What I really want to test for is:
1) if each field has a corresponding `<FormErrors {...{ name}} />`
2) if each field control calls `formControlAttrs` or `formSelectAttrs` with their respective `name`s

I tried mocking what's returned by `useCommonForm` inside each form and had trouble testing it that way with mock functions / a mock JSX element but I'll revisit this - otherwise we need that test span for L114, and on L115, technically you could manually add `name` to the input and `screen.queryByTestId(name)).toBeInTheDocument()` would pass without `formControlAttrs`.

File: app/experimenter/nimbus-ui/src/components/FormAudience/index.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/components/FormAudience/index.tsx:83]: Haven't looked too far into this PR yet, but: Any reason not to fully wrap and pass through the `useForm` hook from within `useCommonForm`?

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/FormAudience/index.tsx:83]: Ooo, that could be a good idea especially because calling `useCommonForm` takes in so many required parameters as-is from `useForm`. Will investigate😈

File: app/experimenter/nimbus-ui/src/lib/getConfigLabel.ts
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/lib/getConfigLabel.ts:1]: Ah sorry I made this PR slightly longer than it needed to be, I pulled this out into its own file because we're already using it in two places and I thought I also needed it on the metrics page. Didn't end up using it since probe sets from the config don't have a `label`, but would like to leave it since we use this in the `Summary` and `TableHighlightsOverview` components.

File: app/experimenter/nimbus-ui/src/components/FormAudience/index.tsx
Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/FormAudience/index.tsx]: Much cleaner this way!

File: app/experimenter/nimbus-ui/src/components/FormAudience/mocks.tsx
Comment (pdehaan) [inline app/experimenter/nimbus-ui/src/components/FormAudience/mocks.tsx:36]: I think CI linting is balking at this line(?):

```
src/components/FormAudience/mocks.tsx(28,10): error TS2741: Property 'isServerValid' is missing in type '{ experiment: getExperiment_experimentBySlug; setSubmitErrors: Dispatch<SetStateAction<Record<string, any>>>; ... 4 more ...; submitErrors: Record<...>; }' but required in type 'FormAudienceProps'.
```

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:68]: Hmm, we wouldn't even need this if we wanted to clear the errors `onChange` of inputs/selects rather than check `onBlur`/`onFocus`. I think that would be better UX anyway since client-side validation updates on change.

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx]: I updated this to clear `onChange` for consistency/simplification. Can revert easily but once it's confirmed we want this `onChange` instead (awaiting response in Slack) I'll squash the 2nd commit/make sure the commit message and tests reflect "onChange" rather than "onBlur".

Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx]: Is this that thing that we were concerned about performance issues with?

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx]: @jodyheavener Yes, I remember us bringing up that `onBlur` would probably be more performant, but this ended up being a lot more simplistic than I thought since it's just an `if` check on if a key/value exists in a small object. Sure it calls the function/check `onChange` but I feel that's fine.

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:31]: I wish there was some way to discover this from the query types....

**Review by jaredlockhart — APPROVED** [review]
This looks really great @LZoog ! So nice to see all this logic consolidated 👏 👏 👏 Great job :D

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx]: Sounds good, this explanation makes sense and I trust this should be fine. We can always review later if something feels off.

File: app/experimenter/nimbus-ui/src/hooks/index.ts
Comment (jodyheavener) [inline app/experimenter/nimbus-ui/src/hooks/index.ts:5]: Just realized this is missing the MPL header, but pretty sure that was me 😅

File: app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx
Comment (jodyheavener) [inline app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx:92]: It looks like across all of these forms `isDirtyUnsaved` is only needed for the `useExitWarning` check. Moving that hook inside your hook could further dry things up.

**Review by jodyheavener — APPROVED** [review]
Tried my hardest but couldn't find anything substantial beyond a suggestion about moving one hook into another. Nice job.

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:37]: Seems like these should live with their respective forms? And FieldNames as a parameterized type at the point of using `useCommonForm`? (e.g. `useCommonForm<OverviewFieldNames>` or suchlike?)

Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:37]: Ooh that seems reasonable and makes it less cumbersome to update should we ever add more forms

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:37]: @lmorchard Hmm yeah, I can move them/look into that. (I was waiting for you to take a look at this PR anyway so feel free to poke more 🙂)

File: app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx
Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx:92]: Hmm, I think I'll file a follow up for this one, have a couple other things to look into. Great call out though.

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:37]: Also, I was curious whether my suggestion was actually harder than it seemed, so I gave it a go as an exercise. I think it ended up being pretty simple. If you haven't already tried it, this seems not to break things:

https://github.com/mozilla/experimenter/compare/issues/4021...lmorchard:issues/4021?expand=1

**Review by lmorchard — APPROVED** [review]
LGTM, nice refactoring! 🥂 

Doesn't touch the branches page, but that will be a harder hill to tackle because it has nested dynamic forms. That could be worth filing a follow up issue - and I would propose someone other than me try doing it so that branch form know-how isn't too concentrated with me

Had one suggestion about further shifting the type names around, but not going to block on that

https://github.com/mozilla/experimenter/compare/issues/4021...lmorchard:issues/4021?expand=1

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx
Comment (k88hudson) [inline app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:75]: Not a blocker but after digging through this and writing some docs, it seems like almost almost no fields should actually be required (the exceptions are `name` for experiment and branch) so it might make sense to reverse this default

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx:75]: @k88hudson Yeah, we brought this up in Slack a day or two ago I think, but we didn't want to tack one more thing onto this since I was already almost done and it'd need more tests etc. I think we already have a ticket for it, if not I'll make another. 👍

**Comment by LZoog** [comment]
@lmorchard Thank you Les for looking into that, that's much better🙌 

Trying to finish a test this morning before sprint close, will merge when I push up latest with Les' fix + that test. Also will link follow up issues here, I think there's 2-3.

File: app/experimenter/nimbus-ui/src/hooks/useCommonForm.test.tsx
Comment (LZoog) [reply app/experimenter/nimbus-ui/src/hooks/useCommonForm.test.tsx:131]: I've attempted this twice now and haven't really gotten anywhere so I made a follow up for it: https://jira.mozilla.com/browse/EXP-780

File: app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx
Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/FormMetrics/index.tsx:102]: Follow up here: #4304

**Pull request merged**
